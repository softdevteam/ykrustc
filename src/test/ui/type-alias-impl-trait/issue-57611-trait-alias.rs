// Regression test for issue #57611
// Ensures that we don't ICE
// FIXME: This should compile, but it currently doesn't

#![feature(trait_alias)]
// revisions: min_tait full_tait
#![feature(min_type_alias_impl_trait)]
#![cfg_attr(full_tait, feature(type_alias_impl_trait))]
//[full_tait]~^ WARN incomplete

trait Foo {
    type Bar: Baz<Self, Self>;

    fn bar(&self) -> Self::Bar;
}

struct X;

impl Foo for X {
    type Bar = impl Baz<Self, Self>;
    //~^ ERROR implementation of `FnOnce` is not general enough
    //~| ERROR implementation of `FnOnce` is not general enough
    //~| ERROR implementation of `FnOnce` is not general enough
    //~| ERROR mismatched types
    //~| ERROR mismatched types

    fn bar(&self) -> Self::Bar {
        |x| x
    }
}

trait Baz<A: ?Sized, B: ?Sized> = Fn(&A) -> &B;

fn main() {}
