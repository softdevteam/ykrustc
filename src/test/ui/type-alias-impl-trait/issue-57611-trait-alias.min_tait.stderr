error: implementation of `FnOnce` is not general enough
  --> $DIR/issue-57611-trait-alias.rs:20:16
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: closure with signature `fn(&'2 X) -> &X` must implement `FnOnce<(&'1 X,)>`, for any lifetime `'1`...
   = note: ...but it actually implements `FnOnce<(&'2 X,)>`, for some specific lifetime `'2`

error[E0308]: mismatched types
  --> $DIR/issue-57611-trait-alias.rs:20:16
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `for<'r> Fn<(&'r X,)>`
              found type `Fn<(&'<empty> X,)>`
note: this closure does not fulfill the lifetime requirements
  --> $DIR/issue-57611-trait-alias.rs:28:9
   |
LL |         |x| x
   |         ^^^^^

error: implementation of `FnOnce` is not general enough
  --> $DIR/issue-57611-trait-alias.rs:20:16
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: closure with signature `fn(&'2 X) -> &'2 X` must implement `FnOnce<(&'1 X,)>`, for any lifetime `'1`...
   = note: ...but it actually implements `FnOnce<(&'2 X,)>`, for some specific lifetime `'2`

error[E0308]: mismatched types
  --> $DIR/issue-57611-trait-alias.rs:20:16
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `for<'r> Fn<(&'r X,)>`
              found type `Fn<(&'<empty> X,)>`
note: this closure does not fulfill the lifetime requirements
  --> $DIR/issue-57611-trait-alias.rs:28:9
   |
LL |         |x| x
   |         ^^^^^

error: implementation of `FnOnce` is not general enough
  --> $DIR/issue-57611-trait-alias.rs:20:16
   |
LL |     type Bar = impl Baz<Self, Self>;
   |                ^^^^^^^^^^^^^^^^^^^^ implementation of `FnOnce` is not general enough
   |
   = note: closure with signature `fn(&'2 X) -> &'2 X` must implement `FnOnce<(&'1 X,)>`, for any lifetime `'1`...
   = note: ...but it actually implements `FnOnce<(&'2 X,)>`, for some specific lifetime `'2`

error: aborting due to 5 previous errors

For more information about this error, try `rustc --explain E0308`.
